File: Realize/6Week/PracticeTree1.py
class BinaryTree:
    def __init__(self, data):
        self.arr = data.copy()

    def pre_order(self, idx):
        if idx >= len(self.arr):
            return
        print(self.arr[idx], end=' ')
        self.pre_order(2 * idx + 1)
        self.pre_order(2 * idx + 2)

    def in_order(self, idx):
        if idx >= len(self.arr):
            return
        self.in_order(2 * idx + 1)
        print(self.arr[idx], end=' ')
        self.in_order(2 * idx + 2)

    def post_order(self, idx):
        if idx >= len(self.arr):
            return
        self.post_order(2 * idx + 1)
        self.post_order(2 * idx + 2)
        print(self.arr[idx], end=' ')

    def level_order(self):
        for c in self.arr:
            print(c, end=' ')

if __name__ == "__main__":
    arr = [chr(ord('A') + i) for i in range(10)]
    bt = BinaryTree(arr)

    print("== Preorder ==")
    bt.pre_order(0)
    print()

    print("== Inorder ==")
    bt.in_order(0)
    print()

    print("== Postorder ==")
    bt.post_order(0)
    print()

    print("== Levelorder ==")
    bt.level_order()
    print()

File: Realize/6Week/PracticeTree1.cpp
#include <iostream>
#include <vector>
using namespace std;

class BinaryTree {
public:
    vector<char> arr;

    BinaryTree(vector<char> data) {
        arr = data;
    }

    void preOrder(int idx) {
        if (idx >= arr.size()) return;
        cout << arr[idx] << " ";
        preOrder(2 * idx + 1);
        preOrder(2 * idx + 2);
    }

    void inOrder(int idx) {
        if (idx >= arr.size()) return;
        inOrder(2 * idx + 1);
        cout << arr[idx] << " ";
        inOrder(2 * idx + 2);
    }

    void postOrder(int idx) {
        if (idx >= arr.size()) return;
        postOrder(2 * idx + 1);
        postOrder(2 * idx + 2);
        cout << arr[idx] << " ";
    }

    void levelOrder() {
        for (char c : arr) {
            cout << c << " ";
        }
    }
};

int main() {
    vector<char> arr(10);
    for (int i = 0; i < 10; i++) {
        arr[i] = 'A' + i;
    }

    BinaryTree bt(arr);

    cout << "== Preorder ==" << endl;
    bt.preOrder(0);
    cout << endl;

    cout << "== Inorder ==" << endl;
    bt.inOrder(0);
    cout << endl;

    cout << "== Postorder ==" << endl;
    bt.postOrder(0);
    cout << endl;

    cout << "== Levelorder ==" << endl;
    bt.levelOrder();
    cout << endl;

    return 0;
}
